drivers/led/led_pwm.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT pwm_leds
+
+/**
+ * @file
+ * @brief PWM driven LEDs
+ */
+
+#include <drivers/led.h>
+#include <drivers/pwm.h>
+#include <device.h>
+#include <zephyr.h>
+#include <sys/math_extras.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(led_pwm);
+
+#define DEV_CFG(dev)	((const struct led_pwm_config *) ((dev)->config))
+#define DEV_DATA(dev)	((struct led_pwm_data *) ((dev)->data))
+
+struct led_pwm {
+	char *pwm_label;
+	uint32_t channel;
+	uint32_t period;
+	pwm_flags_t flags;
+};
+
+struct led_pwm_config {
+	int num_leds;
+	const struct led_pwm *led;
+};
+
+struct led_pwm_data {
+	const struct device *pwm;
+};
+
+static int led_pwm_blink(const struct device *dev, uint32_t led,
+			 uint32_t delay_on, uint32_t delay_off)
+{
+	const struct led_pwm_config *config = DEV_CFG(dev);
+	struct led_pwm_data *data = DEV_DATA(dev);
+	const struct led_pwm *led_pwm;
+	uint32_t period_usec, pulse_usec;
+
+	if (led >= config->num_leds) {
+		return -EINVAL;
+	}
+
+	if (u32_add_overflow(delay_on, delay_off, &period_usec) ||
+	    u32_mul_overflow(period_usec, 1000, &period_usec) ||
+	    u32_mul_overflow(delay_on, 1000, &pulse_usec)) {
+		return -EINVAL;
```

Comment:
`LOG_ERR` may be useful here. A short comment explaining what this does would be also useful IMHO.

Reply:
I added a comment. It is a good idea. But I am not adding a LOG_ERR because this error case is very unlikely and the user should be able to figure out that the delay values he passed are huge :)